consensus/core/src/broadcaster.rs

```diff
@@ -0,0 +1,137 @@
+        _context: Arc<Context>,
+        network_client: Arc<C>,
+        mut rx_block_broadcast: broadcast::Receiver<VerifiedBlock>,
+        peer: AuthorityIndex,
+    ) {
+        const RTT_ESTIMATE_DECAY: f64 = 0.95;
+        const TIMEOUT_THRESHOLD_MULTIPLIER: f64 = 2.0;
+        const TIMEOUT_RTT_INCREMENT_FACTOR: f64 = 1.2;
+
+        let mut rtt_estimate = Duration::from_millis(200);
+        let mut requests = FuturesUnordered::new();
+
+        async fn send_block<C: NetworkClient>(
+            network_client: Arc<C>,
+            peer: AuthorityIndex,
+            rtt_estimate: Duration,
+            block: VerifiedBlock,
+        ) -> (Result<ConsensusResult<()>, Elapsed>, Instant, VerifiedBlock) {
+            let start = Instant::now();
+            let req_timeout = rtt_estimate.mul_f64(TIMEOUT_THRESHOLD_MULTIPLIER);
+            let resp = timeout(
+                req_timeout,
+                network_client.send_block(peer, block.serialized()),
+            )
+            .await;
+            if matches!(resp, Ok(Err(_))) {
+                // Add a delay before retrying.
+                sleep_until(start + req_timeout).await;
+            }
+            (resp, start, block)
+        }
+
+        loop {
+            tokio::select! {
+                result = rx_block_broadcast.recv(), if requests.len() < BROADCAST_CONCURRENCY => {
+                    let block = match result {
+                        Ok(block) => block,
+                        Err(broadcast::error::RecvError::Closed) => {
+                            trace!("Sender to {peer} is shutting down!");
+                            return;
+                        }
+                        Err(broadcast::error::RecvError::Lagged(e)) => {
+                            warn!("Sender to {peer} is lagging! {e}");
+                            // Re-run the loop to receive again.
+                            continue;
+                        }
+                    };
+                    requests.push(send_block(network_client.clone(), peer, rtt_estimate, block));
+                }
+                Some((resp, start, block)) = requests.next() => {
+                    match resp {
+                        Ok(Ok(_)) => {
+                            let now = Instant::now();
+                            rtt_estimate = rtt_estimate.mul_f64(RTT_ESTIMATE_DECAY) + (now - start).mul_f64(1.0 - RTT_ESTIMATE_DECAY);
+                        },
+                        Err(Elapsed { .. }) => {
+                            rtt_estimate = rtt_estimate.mul_f64(TIMEOUT_RTT_INCREMENT_FACTOR);
+                        },
+                        Ok(Err(_)) => {
+                            requests.push(send_block(network_client.clone(), peer, rtt_estimate, block));
+                        },
+                    };
+                }
+            };
+            // Limit RTT estimate to be between 1ms and 10s.
+            rtt_estimate = min(rtt_estimate, Duration::from_secs(10));
```

Comment:
Maybe for now we can also keep some metrics with the estimated rtts per authority? It can be a good enough signal to understand the node communication on the block broadcast level.